authorize/grpc.go

```diff
@@ -58,26 +57,29 @@ func (a *Authorize) Check(ctx context.Context, in *envoy_service_auth_v3.CheckRe
 
 	// take the state lock here so we don't update while evaluating
 	a.stateLock.RLock()
-	reply, err := state.evaluator.Evaluate(ctx, req)
+	res, err := state.evaluator.Evaluate(ctx, req)
 	a.stateLock.RUnlock()
 	if err != nil {
 		log.Error(ctx).Err(err).Msg("error during OPA evaluation")
 		return nil, err
 	}
 	defer func() {
-		a.logAuthorizeCheck(ctx, in, out, reply, u)
+		a.logAuthorizeCheck(ctx, in, out, res, u)
 	}()
 
-	switch {
-	case reply.Status == http.StatusOK:
-		return a.okResponse(reply), nil
-	case reply.Status == http.StatusUnauthorized:
-		if isForwardAuth && hreq.URL.Path == "/verify" {
-			return a.deniedResponse(ctx, in, http.StatusUnauthorized, "Unauthenticated", nil)
-		}
-		return a.redirectResponse(ctx, in)
+	if res.Deny != nil {
+		return a.deniedResponse(ctx, in, int32(res.Deny.Status), res.Deny.Message, nil)
+	}
+
+	if res.Allow {
+		return a.okResponse(res), nil
 	}
-	return a.deniedResponse(ctx, in, int32(reply.Status), reply.Message, nil)
+
+	if isForwardAuth && hreq.URL.Path == "/verify" {
+		return a.deniedResponse(ctx, in, http.StatusUnauthorized, "Unauthenticated", nil)
+	}
+
+	return a.redirectResponse(ctx, in)
```

Comment:
I think this isn't right. We need to make sure we only redirect if the user doesn't have a session.